wagtail/admin/wagtail_hooks.py

```diff
@@ -247,3 +257,203 @@ def register_core_features(features):
+    )
+    features.register_converter_rule('contentstate', 'ul', {
+        'from_database_format': {
+            'ul': ListElementHandler('unordered-list-item'),
+            'li': ListItemElementHandler(),
+        },
+        'to_database_format': {
+            'block_map': {'unordered-list-item': {'element': 'li', 'wrapper': 'ul'}}
+        }
+    })
+    features.register_editor_plugin(
+        'draftail', 'ol', draftail_features.BlockFeature({
+            'type': 'ordered-list-item',
+            'icon': 'list-ol',
+            'description': ugettext('Numbered list'),
+        })
+    )
+    features.register_converter_rule('contentstate', 'ol', {
+        'from_database_format': {
+            'ol': ListElementHandler('ordered-list-item'),
+            'li': ListItemElementHandler(),
+        },
+        'to_database_format': {
+            'block_map': {'ordered-list-item': {'element': 'li', 'wrapper': 'ol'}}
+        }
+    })
+
+    features.register_editor_plugin(
+        'draftail', 'bold', draftail_features.InlineStyleFeature({
+            'type': 'BOLD',
+            'icon': 'bold',
+            'description': ugettext('Bold'),
+        })
+    )
+    features.register_converter_rule('contentstate', 'bold', {
+        'from_database_format': {
+            'b': InlineStyleElementHandler('BOLD'),
+            'strong': InlineStyleElementHandler('BOLD'),
+        },
+        'to_database_format': {
+            'style_map': {'BOLD': 'b'}
+        }
+    })
+    features.register_editor_plugin(
+        'draftail', 'italic', draftail_features.InlineStyleFeature({
+            'type': 'ITALIC',
+            'icon': 'italic',
+            'description': ugettext('Italic'),
+        })
+    )
+    features.register_converter_rule('contentstate', 'italic', {
+        'from_database_format': {
+            'i': InlineStyleElementHandler('ITALIC'),
+            'em': InlineStyleElementHandler('ITALIC'),
+        },
+        'to_database_format': {
+            'style_map': {'ITALIC': 'i'}
+        }
+    })
+
+    features.register_editor_plugin(
+        'draftail', 'link', draftail_features.EntityFeature({
+            'type': 'LINK',
+            'icon': 'link',
+            'description': ugettext('Link'),
+            'source': 'ModalWorkflowSource',
+            'decorator': 'Link',
+            # We want to enforce constraints on which links can be pasted into rich text.
+            # Keep only the attributes Wagtail needs.
+            'attributes': ['url', 'id', 'parentId'],
+            # Keep only links which:
+            # - have an undefined href (only set when pasting), or
+            # - Do not start with # (anchor)
+            # - Do not start with / (internal)
+            # - Do not start with mailto: (email)
+            # For mailto, Wagtail does support those but they could come with query parameters that Wagtail does not support.
+            'whitelist': {
+                'href': '^(?!(#|/|mailto:))',
```

Comment:
Do we want to exclude `tel:` as well?

Reply:
Wouldn't it make more sense for this to be an actual whitelist such as `^(http|https|ftp):` (compare: https://github.com/wagtail/wagtail/blob/9f65e7001f17e922db32628b43028414351fe05f/wagtail/core/whitelist.py#L10), rather than a blacklist pretending to be a whitelist? :-)

Reply:
Yep, `tel` should be excluded as well.

I think I tried to turn this around into a more straightforward whitelist but failed (but maybe I didn't try hard enough). The difficulty here is that we are whitelisting based on the `href` attribute, which is only defined on *pasted* links, and this attribute is removed as part of the paste filtering. If we did `'href': '^http:'`, this would remove all non-pasted links.

It might be possible to do something like `'href': '^(http:|https:|undefined$)'` so it whitelists pasted links while accepting `href` to be undefined for non-pasted links. I'll try again and report back.

Reply:
Yep, I didn't try hard enough. `'href': '^(http:|https:|undefined$)'` is the way to go. Much better, thanks for raising this!